### PR TITLE
[refactor] Move vision parts from processor to model for Gemma3

### PR DESCRIPTION
**Pull request by 2ez4bz**
* Why?

In order to take advantage of (in-flight) batching, we want the heavy, GPU-intensive model operations to happen in the model's forward method, not in the input processor.

* What?

This commit moves the vision tower and multimodal projector out of the input processor and into the model, alongside the language model, for Gemma3.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py [hunk 1]
```diff
@@ -164,20 +148,39 @@ def forward(
...
+
         input_ids, inputs_embeds = fuse_input_embeds(
             embedding_layer=self.llm.model.embed_tokens,
             input_ids=input_ids,
-            mm_embeds=mm_embed,
+            mm_embeds=mm_embeds,
             mm_token_ids=torch.tensor([self.image_token_index
                                        ]).to(input_ids.device))
         logits = self.llm.forward(attn_metadata, input_ids, position_ids,
                                   inputs_embeds, return_context_logits)
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_gemma3vl.py:175]: Please note: I made an update [here](https://github.com/NVIDIA/TensorRT-LLM/pull/5853/) where I pass in an additional `image_token_mask`. If we don't pass that, custom bidirectional mask won't be used for context phase requests.

**Comment by 2ez4bz** [comment]
Actually, I need to investigate some correctness issues.

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py [hunk 2]
```diff
@@ -32,47 +29,17 @@ def __init__(self, model_path, model_config, tokenizer, trust_remote_code):
...
-        processor_output = self.processor(text=text_prompt,
-                                          images=mm_data["image"][0],
-                                          return_dict=True,
-                                          return_tensors="pt",
-                                          device=self.device).to(
-                                              'cuda', dtype=torch.bfloat16)
+        assert "image" in mm_data
+        processor_output = self.processor(
+            text=text_prompt,
+            images=mm_data["image"][0],
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_gemma3vl.py:39]: Dumb question: I know I was using `["image"][0]` myself previously. I see we use [images](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/models/modeling_mistral.py#L244-L248) in Mistral Small. Do you think we should do that here as well?

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py [hunk 3]
```diff
@@ -164,20 +148,39 @@ def forward(
...
+
+    @nvtx_range("[Vision] process")
+    def _get_image_features(self, pixel_values):
+        image_features: Tuple[torch.Tensor] = self.vision_tower(
+            pixel_values).last_hidden_state
+        image_features = self.mm_projector(image_features)
+        return image_features
+
+
+AutoModel.register(Gemma3Config, Gemma3Model)
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma3vl.py]: (re-)Removed.

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py (hunk 1 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma3vl.py]: Yep; rebased and incorporated.

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py (hunk 2 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma3vl.py]: I think best let sleeping dogs lie here 😅

**Comment by 2ez4bz** [comment]
NOTE: the branch for this PR temporarily contains the changes from #5921 .

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py [hunk 4]
```diff
@@ -32,47 +29,29 @@ def __init__(self, model_path, model_config, tokenizer, trust_remote_code):
...
         text_prompt, mm_data = inputs.get("prompt"), inputs.get(
             "multi_modal_data", {})
-        assert 'image' in mm_data
-        processor_output = self.processor(text=text_prompt,
-                                          images=mm_data["image"][0],
-                                          return_dict=True,
-                                          return_tensors="pt",
-                                          device=self.device).to(
-                                              'cuda', dtype=torch.bfloat16)
+        if "image" not in mm_data:
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_gemma3vl.py:36]: Not necessarily true that we always receive image. But, I don't think we have a good way to test the case of text-only input.
Just FYI. Please feel free to resolve.

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py [hunk 5]
```diff
@@ -32,47 +29,29 @@ def __init__(self, model_path, model_config, tokenizer, trust_remote_code):
...
+            raise KeyError("Expected image data in multimodal data for Gemma3.")
+
+        images = mm_data["image"]
+        if len(images) != 1:
```
Comment (brb-nv) [inline tensorrt_llm/_torch/models/modeling_gemma3vl.py:40]: Not necessarily the case. Gemma3 can handle multiple images. I don't know if there's a good way to test this right now.
Just FYI. Please feel free to resolve.

**Review by brb-nv — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py [hunk 6]
```diff
@@ -134,13 +106,37 @@ def __init__(self, model_config: ModelConfig[PretrainedConfig], *args,
...
+
+        vision_tower_weights = filter_weights("vision_tower", weights)
+        missing_keys, _ = self.vision_tower.load_state_dict(
+            vision_tower_weights)
+        if len(missing_keys) > 0:
+            raise KeyError(
+                "Missing the following keys for the vision tower in the checkpoint: "
+                f"[{', '.join(missing_keys)}].")
+
+        projector_weights = filter_weights("multi_modal_projector", weights)
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma3vl.py:134]: Done.

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py (hunk 4 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma3vl.py:36]: Good point - made this `if mm_data and "image" not in mm_data`.

File: tensorrt_llm/_torch/models/modeling_gemma3vl.py (hunk 5 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_gemma3vl.py:40]: Hm, but the input processor will always get batch size one, no? In any case, if I'm accessing `images[0]`, then I need this. Will leave it as is for now.

**Review by QiJune — APPROVED** [review]
